## Diabetes classification using support vector machines

Apply a linear SVM, using the scikit-SVM, for the Pima Indian Women diabetes detection problem on the standardized dataset.

In [2]:
import pandas as pd
from sklearn import svm, grid_search
from sklearn.svm import SVC
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import f1_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
import numpy as np
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [4]:
train = pd.read_csv('diabetes_train-std.csv')
test = pd.read_csv('diabetes_test-std.csv')

X_train = train.drop('classvariable', 1)
y_train = train['classvariable']
X_test = test.drop('classvariable', 1)
y_test = test['classvariable']

In [5]:
#Linear  SVM
params = {'C': [float(i)/1000 for i in range(1,1000,5)]}
print params

{'C': [0.001, 0.006, 0.011, 0.016, 0.021, 0.026, 0.031, 0.036, 0.041, 0.046, 0.051, 0.056, 0.061, 0.066, 0.071, 0.076, 0.081, 0.086, 0.091, 0.096, 0.101, 0.106, 0.111, 0.116, 0.121, 0.126, 0.131, 0.136, 0.141, 0.146, 0.151, 0.156, 0.161, 0.166, 0.171, 0.176, 0.181, 0.186, 0.191, 0.196, 0.201, 0.206, 0.211, 0.216, 0.221, 0.226, 0.231, 0.236, 0.241, 0.246, 0.251, 0.256, 0.261, 0.266, 0.271, 0.276, 0.281, 0.286, 0.291, 0.296, 0.301, 0.306, 0.311, 0.316, 0.321, 0.326, 0.331, 0.336, 0.341, 0.346, 0.351, 0.356, 0.361, 0.366, 0.371, 0.376, 0.381, 0.386, 0.391, 0.396, 0.401, 0.406, 0.411, 0.416, 0.421, 0.426, 0.431, 0.436, 0.441, 0.446, 0.451, 0.456, 0.461, 0.466, 0.471, 0.476, 0.481, 0.486, 0.491, 0.496, 0.501, 0.506, 0.511, 0.516, 0.521, 0.526, 0.531, 0.536, 0.541, 0.546, 0.551, 0.556, 0.561, 0.566, 0.571, 0.576, 0.581, 0.586, 0.591, 0.596, 0.601, 0.606, 0.611, 0.616, 0.621, 0.626, 0.631, 0.636, 0.641, 0.646, 0.651, 0.656, 0.661, 0.666, 0.671, 0.676, 0.681, 0.686, 0.691, 0.696, 0.701, 0.706,

In [6]:
linear_svc = svm.SVC(kernel='linear')
clf = grid_search.GridSearchCV(linear_svc, param_grid = params, cv =10)
linear_model = clf.fit(X_train,y_train)

In [7]:
predict_linear = linear_model.predict(X_test)
pd.DataFrame(confusion_matrix(y_test,predict_linear))

0   1
0  223  22
1   49  63

In [8]:
print "Accuracy: ",accuracy_score(y_test,predict_linear)
print "Mean Error Rate: ",1-accuracy_score(y_test,predict_linear)
print linear_model.best_params_

Accuracy:  0.801120448179
Mean Error Rate:  0.198879551821
{'C': 0.011}


Repeat but now using a Gaussian radial basis kernel.

Listed above are the accuracy, mean error rate and best cost penalty, C, for this classifier. The best C of 0.011 was chosen by 10-fold cross validation.

In [9]:
#Gaussian SVM
params = {'C': [float(i)/100 for i in range(1,1000,5)]}
print params

{'C': [0.01, 0.06, 0.11, 0.16, 0.21, 0.26, 0.31, 0.36, 0.41, 0.46, 0.51, 0.56, 0.61, 0.66, 0.71, 0.76, 0.81, 0.86, 0.91, 0.96, 1.01, 1.06, 1.11, 1.16, 1.21, 1.26, 1.31, 1.36, 1.41, 1.46, 1.51, 1.56, 1.61, 1.66, 1.71, 1.76, 1.81, 1.86, 1.91, 1.96, 2.01, 2.06, 2.11, 2.16, 2.21, 2.26, 2.31, 2.36, 2.41, 2.46, 2.51, 2.56, 2.61, 2.66, 2.71, 2.76, 2.81, 2.86, 2.91, 2.96, 3.01, 3.06, 3.11, 3.16, 3.21, 3.26, 3.31, 3.36, 3.41, 3.46, 3.51, 3.56, 3.61, 3.66, 3.71, 3.76, 3.81, 3.86, 3.91, 3.96, 4.01, 4.06, 4.11, 4.16, 4.21, 4.26, 4.31, 4.36, 4.41, 4.46, 4.51, 4.56, 4.61, 4.66, 4.71, 4.76, 4.81, 4.86, 4.91, 4.96, 5.01, 5.06, 5.11, 5.16, 5.21, 5.26, 5.31, 5.36, 5.41, 5.46, 5.51, 5.56, 5.61, 5.66, 5.71, 5.76, 5.81, 5.86, 5.91, 5.96, 6.01, 6.06, 6.11, 6.16, 6.21, 6.26, 6.31, 6.36, 6.41, 6.46, 6.51, 6.56, 6.61, 6.66, 6.71, 6.76, 6.81, 6.86, 6.91, 6.96, 7.01, 7.06, 7.11, 7.16, 7.21, 7.26, 7.31, 7.36, 7.41, 7.46, 7.51, 7.56, 7.61, 7.66, 7.71, 7.76, 7.81, 7.86, 7.91, 7.96, 8.01, 8.06, 8.11, 8.16, 8.21, 8.2

In [10]:
gaussian_svc = svm.SVC(kernel='rbf')
clf = grid_search.GridSearchCV(gaussian_svc, param_grid = params, cv =10)
gaussian_model = clf.fit(X_train,y_train)

In [11]:
predict_gaussian = gaussian_model.predict(X_test)
pd.DataFrame(confusion_matrix(y_test,predict_gaussian))

0   1
0  226  19
1   50  62

In [12]:
print "Accuracy: ",accuracy_score(y_test,predict_gaussian)
print "Mean Error Rate: ",1-accuracy_score(y_test,predict_gaussian)
print gaussian_model.best_params_

Accuracy:  0.806722689076
Mean Error Rate:  0.193277310924
{'C': 0.16}


Listed above are the accuracy, mean error rate and best cost penalty, C, for this classifier. The best C of 0.16 was chosen by 10-fold cross validation.